service/backend/pchain/account.go

```diff
@@ -117,38 +117,56 @@ func (b *Backend) AccountCoins(ctx context.Context, req *types.AccountCoinsReque
 	}
 	fetchSharedMemory := subAccountAddress == pmapper.SubAccountTypeSharedMemory
 
-	height, utxos, _, typedErr := b.fetchUTXOsAndStakedOutputs(ctx, addr, false, fetchSharedMemory, nil)
+	// utxos from fetchUTXOsAndStakedOutputs are guarateed to:
+	// 1. be unique (no duplicates)
+	// 2. containt only assetIDs
+	// 3. have not multisign utxos
+	// by parseAndFilterUTXOs call in fetchUTXOsAndStakedOutputs
+	height, utxos, _, typedErr := b.fetchUTXOsAndStakedOutputs(ctx, addr, false, fetchSharedMemory, assetIDs)
 	if typedErr != nil {
 		return nil, typedErr
 	}
 
-	// convert raw UTXO bytes to Rosetta Coins
-	coins, err := b.processUtxos(currencyAssetIDs, utxos)
-	if err != nil {
-		return nil, service.WrapError(service.ErrInternalError, err)
+	// convert UTXOs to Rosetta Coins
+	coins := []*types.Coin{}
+	for _, utxo := range utxos {
+		amounter, ok := utxo.Out.(avax.Amounter)
+		if !ok {
+			return nil, service.WrapError(service.ErrInternalError, errUnableToGetUTXOOut)
+		}
+		coin := &types.Coin{
+			CoinIdentifier: &types.CoinIdentifier{Identifier: utxo.UTXOID.String()},
+			Amount: &types.Amount{
+				Value:    strconv.FormatUint(amounter.Amount(), 10),
+				Currency: mapper.AtomicAvaxCurrency,
+			},
+		}
+		coins = append(coins, coin)
 	}
 
 	block, err := b.indexerParser.ParseNonGenesisBlock(ctx, "", height)
 	if err != nil {
 		return nil, service.WrapError(service.ErrInvalidInput, "unable to get height")
 	}
 
+	// this is needed just for sorting. Uniqueness is guaranteed by utxos uniqueness
+	coins = common.SortUnique(coins)
 	return &types.AccountCoinsResponse{
 		BlockIdentifier: &types.BlockIdentifier{
 			Index: int64(height),
 			Hash:  block.BlockID.String(),
 		},
-		Coins: common.SortUnique(coins),
+		Coins: coins,
 	}, nil
 }
 
-func (b *Backend) fetchBalance(ctx context.Context, addrString string, fetchImportable bool, currencyAssetIds map[ids.ID]struct{}) (uint64, *AccountBalance, *types.Error) {
+func (b *Backend) fetchBalance(ctx context.Context, addrString string, fetchImportable bool, assetIds ids.Set) (uint64, *AccountBalance, *types.Error) {
```

Comment:
ids.Set is basically an alias for map[ids.ID]struct{} with some convenience methods along it.